### DISCUSSION TITLE
About "Saving time is too long"

### DISCUSSION THREAD

**Discussion by johnsoncodehk** [vuejs/language-tools#2740] [General]
VSCode update sometimes change the scheduling strategy of codeActions and formatting execution, which may cause performance problems when users install a specific extension combination and enable `editor.codeActionsOnSave`.

The reason behind this may be that VSCode emits document changes events multiple times during a save cycle, which causes extensions based on TypeScript LanguageService to perform expensive TS state synchronization multiple times. However, I haven't actually encountered this problem, so this is just a wild guess.

For this problem, we have nothing can do in Volar, so after v1.4.0, we just prompt you to disable Volar's codeActions when we detect that the storage time is too long.

**Comment by WhiteCoffee9834** [comment]


**Reply by ishowman** [reply]
希望在安装插件的文档中就说明下这个问题，不然大家遇到了都一脸懵逼

**Comment by gyhyfj** [comment]
Very serious problem, especially for big project, cannot save code change in 1 mins somtimes on windows

**Reply by yanglianjin** [reply]
yeah...

**Comment by laygir** [comment]
Having these issues in a very small project with eslint+prettier configs/plugins 😞

**Reply by BlackNoar** [reply]
yeah

**Reply by xuiang** [reply]
when I encounter this problem, I will manually close 'codeActions' at that time

**Reply by xuiang** [reply]
else waste a lot of time waiting

**Comment by or2e** [comment]
Also have problems in a [small project](https://github.com/verdie-g/crpg/tree/master/src/WebUI)
After 10-20 minutes of work, the file saving time increases to 10-30 seconds

**Comment by V35CH** [comment]
I also have the problem.
I have to kill the "Code Helper (Plugin)" process. Which is >= 100%. Then vs code works as expected for 5min. Then the CPU raises up again.

**Comment by seanogdev** [comment]
Does anyone else here have the prettier-tailwind plugin installed? We've seen major slowdowns with 0.2.6+ https://github.com/tailwindlabs/prettier-plugin-tailwindcss/issues/144

**Reply by gyhyfj** [reply]
Yes, it will showdown as long as there's plugin doing something on save.

**Reply by Charith47** [reply]
I have this plugin installed. Facing the same issue. Also seem to cause very high disk usage. I'm on WSL 2

**Reply by seanogdev** [reply]
We've downgraded to v0.2.5 for now until https://github.com/tailwindlabs/prettier-plugin-tailwindcss/issues/168 gets resolved @Charith47

**Comment by noopurphalak** [comment]
How can I disable this message?

**Reply by joezimjs** [reply]
Just set the timeout value to a very large number

**Reply by delljin** [reply]
can you tell me how to set it?

**Reply by delljin** [reply]
if i set it,would it can't save

**Reply by joezimjs** [reply]
vue.codeActions.savingTimeLimit is the setting ID. Or if you're using the settings GUI:
Extensions > Volar > "Vue > Code Actions: Saving Time Limit"

Just set to really high number (e.g 999999) and it'll stop bugging you during saves... but the saves won't happen any faster.

**Comment by jose-azevedo** [comment]
I'm also having this problem. Sometimes it takes more than a minute just to save a file.

**Reply by kewinzaq1** [reply]
same

**Comment by gyhyfj** [comment]
Is there any plan to resolve this issue by us or vscode?

**Reply by delljin** [reply]
change you IDE

**Comment by mlesin** [comment]
Disabling this option fixed vscode slowdowns for me:

`noErrorTruncation: true`

https://github.com/vuejs/language-tools/issues/2533#issuecomment-1543496140

**Comment by chrisbbreuer** [comment]
I am experiencing the same issues in [Stacks.js](https://github.com/stacksjs/stacks) currently, too. Will see within the next few days if I can find a way to resolve the issues for myself there. Will report back here, if I do 🙂

_Stacks is very similar to Vitesse, in several ways. It uses Vue as its UI engine, UnoCSS as its CSS engine, Vite & unbuild as its build system, and a whole host of other well-known community packages, to form a sophisticated foundation for the upcoming Stacks ecosystem._

_The code base is not anywhere near stable yet, so I am not going to rule out any other issues, but it did start degrading heavily in performance within the last month. Even though I upgraded around that time to an M2 Pro, my M1 Air was performing much better last month—which led me to deep-dive profiling today until I found this thread_

**Reply by chrisbbreuer** [reply]
I haven't had any luck so far. For now, temporarily, we are shifting to create WebStorm support because the performance degradation is too stark. _The good news: Volar is working well_.

Will come back to this within the next few weeks.

**Comment by laygir** [comment]
I am under time constraints in my project so didn't have time to dig this further and I had to disable much needed linting/formatting tools to get rid of this minutes of wait times, never had such behaviour with an extension before :/

**Comment by tycrimm** [comment]
Disgustingly brute force but reboots everything long enough to snap my plugins out of their drunken stupor:

`ps aux | grep "Code Helper (Plugin)" | awk '{print $2}' | xargs kill`

Tie this in with `watch` and you almost never have to think about it again 🤦‍♂️

**Comment by moHaHa** [comment]
any solutions?  :/

**Comment by moHaHa** [comment]
any solutions?  :/

**Comment by francisashley** [comment]
hellooo 🙏🙏🙏

**Comment by johnsoncodehk** [comment]
For people asking for a fix/complainers: We've probably had over 10 related issues until now, but none of them provided a reproduction, and we couldn't even investigate it. We can't do more than prompt to disable code actions.

Honestly I'm exhausted from this, please provide a reproduction, or shut up.

**Comment by qhl123456** [comment]
same，When the project is large, it can be slow to save

**Comment by gyhyfj** [comment]
vscode team want people to abandon vue maybe

**Reply by johnsoncodehk** [reply]
What do you mean? This is not a Vue-specific problem.

**Comment by joezimjs** [comment]
The only way I've been able to get this sped up at this point is to disable the vue codeActions. I'm curious if someone can quickly explain what these codeActions are doing? My code still gets fixed by eslint super quickly on save, so am I actually missing out on anything?

**Reply by johnsoncodehk** [reply]
We forwards the code action request to internal ts / html / css... language services.

Is it still slow if you disable ESLint instead of Volar codeCctions?

**Reply by joezimjs** [reply]
If I disable ESLint codeActions and enable Volar codeActions, it's no longer sluggish (assuming I don't add the `source.fixAll` or `source.fixAll.eslint` or `source.fixAll.volar` codeAction to the list of `codeActionsOnSave`), but Volar isn't fixing the linting issues, or doing anything as far as I can detect but ESLint is. 

What kinds of codeActions do those language services perform? The language services are running and reporting errors as I type, right? So what do they do on save? (You can just point to docs if that's easier, but I can't find anything)

**Reply by akowalz** [reply]
+1 - The issue was fixed for me by configuring `"vue.codeActions.enabled": false"` in my workspace settings, and Vue-specific fixes like reordering of v- directives are still happening on save, even with that disabled. Is there any downside to turning `vue.codeActions` off?

**Reply by yp910108** [reply]
After setting `"vue.codeActions.enabled": false"`, the quick fix was lost when I pressed `⌘.`. Is there a problem with my testing?

**Reply by so1ve** [reply]
This is intended. This setting disables quick fix. Upgrading volar to latest version, you'll see a performance bump.

**Comment by stanzakharin** [comment]
For me, the issue ended up being that I had the extensions **Vue Language Features (Volar)** _and_ **TypeScript Vue Plugin (Volar)** installed at the same time. As soon as I disabled the TypeScript extension, my saves went back to normal. I'm still able to use TypeScript without any issues, so I'm honestly not sure why the [official documentation](https://vuejs.org/guide/typescript/overview.html#ide-support) states that you need both.

**Reply by joezimjs** [reply]
I tried this and it didn't help. If you need the TS Vue one, you still need the Vue one, right?

**Comment by akowalz** [comment]
Is there a place where I can find out what code actions this plugin is providing? I've found that disabling the plugin's code actions with `"vue.codeActions.enabled": false` fixes this issue and makes saving snappy again in my setup. And even with this disabled, some Vue-specific fixes like reordering of `v-` directives on components still get applied on save (presumably by ESLint?). For now, this seems like the best solution for me, but I'd like to know what other fixes I may be missing out on. Is this documented anywhere?

**Reply by joezimjs** [reply]
I [asked the same thing](https://github.com/vuejs/language-tools/discussions/2740#discussioncomment-6255138) (just two questions above this one) and got [this answer](https://github.com/vuejs/language-tools/discussions/2740#discussioncomment-6267095). Not a super helpful answer, but then again, they can't answer much more deeply because it's not their plugin doing the work.

**Comment by JackieDYH** [comment]
Saving time is too long.(1040 ms 1000 ms),
Disable codeActions
Increase saveTimeLimit
What is this?

保存时间太长，怎么办

**Comment by johnsoncodehk** [comment]
Someone told me that the problem no longer occurs after upgrading to VSCode 1.81.0-insider, if you still have this problem please take a try with 1.81.0-insider.

**Reply by joezimjs** [reply]
With 1.81 released now, I've tried it. Right now, instead of 5+ seconds, it's doing it in <1 second. Huge gain! But, I still see a very visible and noticeable delay between hitting save and code fixes appearing in the editor, which means it's taking half a second or more. This is definitely fine right now, but it seems to indicate the possibility of it getting longer as the app grows.

**Comment by ccxdev** [comment]
Any updates ?)

**Comment by Akryum** [comment]
https://github.com/volarjs/volar.js/pull/66 has been released in [v1.8.16](https://github.com/vuejs/language-tools/blob/master/CHANGELOG.md#1816-2023107), please give it a try and see if there are still performance issues.

**Reply by vidal7** [reply]
Wow! From our about 1400 components monolith code base, saving a vue file is now nearly instantaneous. In 1.8.15, saving a vue file took sometime 30 seconds to a full minute! Very happy about it thank you!

Also I noticed another very appreciable performance gain.

Getting type information on a child component takes 26 seconds instead of 56 seconds when restarting vscode. A performance gain of 54%!

Great job!

**Reply by joezimjs** [reply]
So far seems much better.
